allennlp/modules/seq2seq_encoders/efficient_multi_head_attention.py

```diff
@@ -0,0 +1,159 @@
+        if mask is None:
+            mask = Variable(inputs.data.new(batch_size, timesteps).fill_(1.0))
+
+        # Treat the queries, keys and values each as a ``num_heads`` size batch.
+        # shape (num_heads, batch_size * timesteps, hidden_dim)
+        inputs_per_head = inputs.repeat(num_heads, 1, 1).view(num_heads,
+                                                              batch_size * timesteps,
+                                                              hidden_dim)
+        # Do the projections for all the heads at once.
+        # Then reshape the result as though it had a
+        # (num_heads * batch_size) sized batch.
+        queries_per_head = torch.bmm(inputs_per_head, self._query_projections)
+        # shape (num_heads * batch_size, timesteps, attention_dim)
+        queries_per_head = queries_per_head.view(num_heads * batch_size,
+                                                 timesteps,
+                                                 self._attention_dim)
+
+        keys_per_head = torch.bmm(inputs_per_head, self._key_projections)
+        # shape (num_heads * batch_size, timesteps, attention_dim)
+        keys_per_head = keys_per_head.view(num_heads * batch_size,
+                                           timesteps,
+                                           self._attention_dim)
+
+        values_per_head = torch.bmm(inputs_per_head, self._value_projections)
+        # shape (num_heads * batch_size, timesteps, attention_dim)
+        values_per_head = values_per_head.view(num_heads * batch_size, timesteps, self._values_dim)
+
+        # shape (num_heads * batch_size, timesteps, timesteps)
+        scaled_similarities = torch.bmm(queries_per_head, keys_per_head.transpose(1, 2)) / self._scale
+
+        # shape (num_heads * batch_size, timesteps, timesteps)
+        # Normalise the distributions, using the same mask for all heads.
+        attention = last_dim_softmax(scaled_similarities, mask.repeat(num_heads, 1))
+        attention = self._attention_dropout(attention)
+        # This is doing the following batch-wise matrix multiplication:
+        # (num_heads * batch_size, timesteps, timesteps) *
+        # (num_heads * batch_size, timesteps, values_dim)
+        # which is equivalent to a weighted sum of the values with respect to
+        # the attention distributions for each element in the num_heads * batch_size
+        # dimension.
+        # shape (num_heads * batch_size, timesteps, values_dim)
+        outputs = torch.bmm(attention, values_per_head)
+
+        # Reshape back to original shape (batch_size, timesteps, num_heads * values_dim)
+        # Note that we _cannot_ use a reshape here, because this tensor was created
+        # with num_heads being the first dimension, so reshaping naively would not
+        # throw an error, but give an incorrect result.
+        # TODO(Mark): Given the possible slowness of split, maybe reshape + transpose is better?
```

Comment:
Yeah, my intuition says that a reshape + transpose should be a lot faster than a split + cat.  Can you run some simple timing tests to see if it makes a difference?

Reply:
I tried this, and actually it's not slower, because to get it in the right shape for a `view` I need to `view, contiguous, permute, contiguous and view` it.

Reply:
Interesting, that's good to know.  Thanks again for checking.  It'll help us get a bit of a better intuition for which operations we should be afraid of.